src/db/gc/mod.rs

```diff
@@ -0,0 +1,435 @@
+use crate::chain::ChainEpochDelta;
+
+use crate::db::{truncated_hash, GarbageCollectable};
+use crate::ipld::stream_graph;
+use crate::shim::clock::ChainEpoch;
+use ahash::{HashSet, HashSetExt};
+use futures::StreamExt;
+use fvm_ipld_blockstore::Blockstore;
+use std::mem;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time;
+use tracing::info;
+
+/// [`MarkAndSweep`] is a simple garbage collector implementation that traverses all the database
+/// keys writing them to a [`HashSet`], then filters out those that need to be kept and schedules
+/// the rest for removal.
+///
+/// Note: The GC does not know anything about the hybrid CAR-backed and ParityDB approach, only
+/// taking care of the latter.
+pub struct MarkAndSweep<DB> {
+    db: Arc<DB>,
+    get_heaviest_tipset: Box<dyn Fn() -> Arc<Tipset> + Send>,
+    marked: HashSet<u32>,
+    epoch_marked: ChainEpoch,
+    depth: ChainEpochDelta,
+    block_time: Duration,
+}
+
+impl<DB: Blockstore + GarbageCollectable> MarkAndSweep<DB> {
+    /// Creates a new mark-and-sweep garbage collector.
+    ///
+    /// # Arguments
+    ///
+    /// * `db` - A reference to the database instance.
+    /// * `get_heaviest_tipset` - A function that facilitates heaviest tipset retrieval.
+    /// * `depth` - The number of state-roots to retain.
+    /// * `block_time` - An average block production time.
+    pub fn new(
+        db: Arc<DB>,
+        get_heaviest_tipset: Box<dyn Fn() -> Arc<Tipset> + Send>,
+        depth: ChainEpochDelta,
+        block_time: Duration,
+    ) -> Self {
+        Self {
+            db,
+            get_heaviest_tipset,
+            depth,
+            marked: HashSet::new(),
+            epoch_marked: 0,
+            block_time,
+        }
+    }
+    // Populate the initial set with all the available database keys.
+    fn populate(&mut self) -> anyhow::Result<()> {
+        self.marked = self.db.get_keys()?;
+        Ok(())
+    }
+
+    // Filter out the initial set, leaving only the entries that need to be removed.
+    // NOTE: One concern here is that this is going to consume a lot of CPU.
+    async fn filter(&mut self, tipset: Arc<Tipset>, depth: ChainEpochDelta) -> anyhow::Result<()> {
+        // NOTE: We want to keep all the block headers from genesis to heaviest tipset epoch.
+        let mut stream = stream_graph(&self.db, (*tipset).clone().chain(&self.db), depth);
+
+        while let Some(block) = stream.next().await {
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let marked = mem::take(&mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
```

Comment:
We only have an `interval` argument here, but you are mentioning a `depth` argument as well.

Reply:
yep, outdated docs. Good spot!